### PR TITLE
FCFs: Deprecate 'lambda' keyword and 'func' type constructor

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
This PR deprecates use of the 'lambda' keyword as well as the legacy
'func' type constructor.

```chapel
const p = lambda() { writeln('Deprecated!'); }
type T = func(int, int, int); // Deprecated!
```

Users should prefer `proc` syntax instead when constructing anonymous
procedures and procedure types.

```chapel
const p = proc() { writeln('OK!'); }
type T = proc(_: int, _: int): int;
```

Note that this deprecation will have an impact on type computations.
Currently, procedure types constructed with `proc` syntax include
the names of formal arguments in the resulting procedure type. These
names can be anonymous `_`, but cannot be omitted.

Implicit conversion from named (e.g., `x`) to anonymous `_` formals
has not been implemented yet. This means that users cannot yet write:

```chapel
const p1 = proc(x: int, y: int) { return x + y; };
const p2: proc(_: int, _: int): int = p1; // Cannot implicitly convert.
```

Implicit conversion rules will be considered and implemented at a
later date. The current restrictions may cause some friction when
migrating deprecated code, so be aware (this is largely only a
problem if the `func` type constructor is used, which I highly doubt
we will see in the wild).

While here, mark the `argTypes` and `retType` methods on FCFs as
unstable. I suspect we will want to adopt a different strategy
for reasoning about function types in the future.

---

Remove the `fcfsUseLegacyBehavior` flag now that legacy behaviors
have been deprecated. Now only the `fcfsUsePointerImplementation`
flag remains.

Remove many tests exercising `lambda` that already have equivalents
in `test/functions/fcf`.

Adjust resolution of function capture to error for unsupported return
intents (forgot to move old checks out of lowering functions).

Collapse many "higher order function spectests" into a single test
named `HigherOrderPatterns.chpl`. These may be useful to keep
around for when closures are implemented.

Add futures documenting bugs in the current FCF implementation.

Reviewed by @jeremiah-corrado. Thanks!

TESTING

- [x] `linux64`, `standard`

### CODE REVIEW

File: frontend/lib/uast/post-parse-checks.cpp [hunk 1]
```diff
@@ -780,6 +782,12 @@ void Visitor::checkNoReceiverClauseOnPrimaryMethod(const Function* node) {
   }
 }
 
+void Visitor::checkLambdaDeprecated(const Function* node) {
+  if (node->kind() != Function::LAMBDA) return;
+  warn(node, "'lambda' syntax is deprecated, please construct anonymous "
+             "procedures using the 'proc' keyword instead");
+}
+
```
Comment (DanilaFe) [inline frontend/lib/uast/post-parse-checks.cpp:790]: Can't you use the `Deprecation` error type here?

Comment (dlongnecke-cray) [reply frontend/lib/uast/post-parse-checks.cpp:790]: Right now that requires information that can't be computed for this such as target IDs. I'd have to go through and refactor it. In general we probably need a slightly different strategy for deprecating language features.

File: test/functions/lambda/lambdaCapturing.good [hunk 2]
```diff
@@ -1,3 +0,0 @@
-lambdaCapturing.chpl:5: In function 'test':
```
Comment (jeremiah-corrado) [inline test/functions/lambda/lambdaCapturing.good:1]: It looks like you deleted the good file, but not the source file here.

File: test/functions/lambda/lambdaThrows.chpl [hunk 3]
```diff
@@ -1,8 +0,0 @@
-const f1 = lambda(n: int): int throws { return n*n; };
-const f2 = lambda(n: int): int throws { throw new Error(n:string); };
```
Comment (jeremiah-corrado) [inline test/functions/lambda/lambdaThrows.chpl:2]: Can the new Fcfs throw?

**Review by jeremiah-corrado — APPROVED** [review]
Looks good! Just a few minor notes:

File: test/functions/jturner/lambda_ref_ret_type.good [hunk 4]
```diff
@@ -1 +0,0 @@
-lambda_ref_ret_type.chpl:3: error: 'ref' return types are not allowed in lambdas
```
[the remark this answers is not in the document]
Comment (dlongnecke-cray) [reply test/functions/jturner/lambda_ref_ret_type.good:1]: Yeah there is a test in `functions/fcf/Motivators.chpl` for that already.

File: test/functions/jturner/fun_type_annot_void.chpl [hunk 5]
```diff
@@ -1,8 +1,8 @@
 use Functions;
 
-var g : func(void) = lambda() { writeln("hello, world!"); };
+var g : proc() = proc() { writeln("hello, world!"); };
 g();
 
-var h : func() = lambda() { writeln("hello, world!"); };
+var h : proc() = proc() { writeln("hello, world!"); };
```
[the remark this answers is not in the document]
Comment (dlongnecke-cray) [reply test/functions/jturner/fun_type_annot_void.chpl:6]: Just removed the whole test.

File: test/functions/lambda/lambdaCapturing.good (hunk 2 above)
Comment (dlongnecke-cray) [reply test/functions/lambda/lambdaCapturing.good:1]: I'm not seeing that file now so I must have removed it at some point.

File: test/functions/lambda/lambdaThrows.chpl (hunk 3 above)
Comment (dlongnecke-cray) [reply test/functions/lambda/lambdaThrows.chpl:2]: Yes they can, and there is a test for that in `functions/fcf/Motivators.chpl` that mirrors this one.

**Review by jeremiah-corrado — APPROVED** [review]

**Pull request merged**